ruby/lib/google/protobuf/repeated_field.rb

```diff
@@ -29,11 +29,141 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 # add syntatic sugar on top of the core library
+require 'forwardable'
 module Google
   module Protobuf
     class RepeatedField
+      extend Forwardable
 
-      alias_method :size, :length
+      # methods defined in C or Java:
+      #   +
+      #   [], at
+      #   []=
+      #   clear
+      #   dup, clone
+      #   each
+      #   each_with_index
+      #   hash
+      #   insert
+      #   push, <<
+      #   pop
+      #   replace
+      #   length, size
+      #   ==
+      #   to_ary, to_a
+      #   also all enumerable
+
+      # NOTE:  using delegators rather than method_missing to make the
+      #        relationship explicit instead of implicit
+      # NOTE2: use to_a instea
+      def_delegators :to_ary,
+        :&, :*, :-, :'<=>',
+        :assoc, :bsearch, :combination, :compact, :count, :cycle,
+        :drop, :drop_while, :eql?, :fetch, :find_index, :flatten,
+        :include?, :index, :inspect, :join,
+        :pack, :permutation, :product, :pretty_print, :pretty_print_cycle,
+        :rassoc, :repeated_combination, :repeated_permutation, :reverse, :rindex, :rotate,
+        :sample, :shelljoin, :slice, :to_s, :transpose, :uniq, :|
+
+      def_delegators :to_a, :shuffle
+
+
+      def first(n=nil)
+        n ? self[0..n] : self[0]
+      end
+
+
+      def last(n=nil)
+        n ? self[(self.size-n-1)..-1] : self[-1]
+      end
+
+
+      def pop(n=nil)
+        if n
+          results = []
+          n.times{ results << pop_internal }
+          return results
+        else
+          return pop_internal
+        end
+      end
+
+
+      def empty?
+        self.size == 0
+      end
+
+      # array aliases into enumerable
+      alias_method :each_index, :each_with_index
+      alias_method :slice, :[]
+      alias_method :values_at, :select
+      alias_method :map, :collect
+
+
+      %w(delete_at delete_if shift slice! unshift).each do |method_name|
```

Comment:
the self-modifying methods in `Array` can be consolidated into these three patterns.